java/ql/src/experimental/CWE-939/IncorrectURLVerification.ql

```diff
@@ -0,0 +1,64 @@
+/**
+ * @id java/incorrect-url-verification
+ * @name Insertion of sensitive information into log files
+ * @description Apps that rely on URL parsing to verify that a given URL is pointing to a trusted server are susceptible to wrong ways of URL parsing and verification. 
+ * @kind problem
+ * @tags security
+ *       external/cwe-939
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+import DataFlow
+import PathGraph
+
+
+/**
+ * The Java class `android.net.Uri` and `java.net.URL`.
+ */
+class Uri extends RefType {
+    Uri() {
+        hasQualifiedName("android.net", "Uri") or
+        hasQualifiedName("java.net", "URL")
+    }
+}
+
+/**
+ * The method `getHost()` declared in `android.net.Uri` and `java.net.URL`.
+ */
+class UriGetHostMethod extends Method {
+    UriGetHostMethod() {
+        getDeclaringType() instanceof Uri and
+        hasName("getHost") and
+        getNumberOfParameters() = 0
+    }
+}
+
+/**
+ * A library method that acts like `String.format` by formatting a number of
+ * its arguments according to a format string.
+ */
+class HostVerificationMethodAccess extends MethodAccess {
+    HostVerificationMethodAccess() {
+        (
+
+            this.getMethod().hasName("endsWith") or
+            this.getMethod().hasName("contains") or
+            this.getMethod().hasName("indexOf")
+        ) and
+        this.getMethod().getNumberOfParameters() = 1 and 
+        (
+            this.getArgument(0).(StringLiteral).getRepresentedString().charAt(0) != "." or  //string constant comparison
+            this.getArgument(0).(AddExpr).getLeftOperand().(VarAccess).getVariable().getAnAssignedValue().(StringLiteral).getRepresentedString().charAt(0) != "." or   //var1+var2, check var1 starts with "."
+            this.getArgument(0).(AddExpr).getLeftOperand().(StringLiteral).getRepresentedString().charAt(0) != "." or  //"."+var2, check string constant "."
+            exists (MethodAccess ma | this.getArgument(0) = ma and ma.getMethod().hasName("getString") and ma.getArgument(0).toString().indexOf("R.string") = 0) or  //res.getString(R.string.key)
```

Comment:
I'm not sure exactly what this line is trying to match.  But in any case, it's bad practise to rely on the `.toString()` representation when you're matching some piece of syntax.

Reply:
Hi @aschackmull ,

Thanks for reviewing this PR.

For those lines, I've added comments to all of them but they may be not very clear.
+            this.getArgument(0).(StringLiteral).getRepresentedString().charAt(0) != "." or  //string constant comparison e.g. uri.getHost().endsWith("example.com")
+            this.getArgument(0).(AddExpr).getLeftOperand().(VarAccess).getVariable().getAnAssignedValue().(StringLiteral).getRepresentedString().charAt(0) != "." or   //var1+var2, check var1 starts with "." e.g. String domainName = "example"; Uri.parse(url).getHost().endsWith(domainName+".com")
+            this.getArgument(0).(AddExpr).getLeftOperand().(StringLiteral).getRepresentedString().charAt(0) != "." or  //"."+var2, check string constant "." e.g. String domainName = "example.com";  Uri.parse(url).getHost().endsWith("www."+domainName)
+            exists (MethodAccess ma | this.getArgument(0) = ma and ma.getMethod().hasName("getString") and ma.getArgument(0).toString().indexOf("R.string") = 0) or  //Check resource properties in /res/values/strings.xml in Android mobile applications using res.getString(R.string.key)

So basically I've tried to make the query as comprehensive as possible so that it can detect as many patterns as possible. The following five patterns are checked:
+ String literal e.g. "example.com"
+ Variable concatenated with a literal suffix e.g. domainName+".com"
+ Variable concatenated with a literal prefix e.g. "www."+domainName
+ String literal defined in Android property files although it could be false positive sometimes since /res/values/strings.xml for Java Android projects are not included in the database yet e.g. res.getString(R.string.domainName)
+ Variable e.g. domainName

Hope this clarifies what I'm trying to achieve. Please let me know if there are ways to optimize.

Thanks,
@luchua-bc

Reply:
I also formatted the ql file with auto-format. Would you please check whether the code passes PR checks now? I will use the same approach to format the code if it does.

Thanks,
@luchua-bc